### ISSUE TITLE
Install Strapi v5 in Sub-Path on Server not working

### ISSUE THREAD

**Issue by JWandscheer** [strapi/strapi#21184]
### Required System information

- Node.js version: v18
- Strapi version: 5.0.0-rc.17
- Database: postgres
- Operating system: Linux
- Is your project Javascript or Typescript: Typescript

### Describe the bug

Installing Strapi in a Sub-Path on a Server (f.e. https://rooturl.tld/backend) seems nearly unpossible. In a fresh install with server config url: https://rooturl.tld/backend/ and admin config admin-url: /admin (at build-time and runtime) the admin page generally loads (http 200) in /backend/admin, but the resources are loaded from the root path (https://rooturl.tld/strapi-BA4UeQmR.js). This leads to 404 and a blank page.

### Steps to reproduce the behavior

1. Build Strapi with ENV URL=https:/rooturl.tld/backend/, ENV ADMIN_URL=/admin
2. Run Container
3. Admin Page is blank because js resources could not be loaded

### Expected behavior

The resources should not ignore server-paths.

### Code snippets

We are running strapi in a kubernetes cluster and have a rewrite in our virtual service:
```
- name: "backend-routes"
    match:
    - uri:
        prefix: "/backend/"
    rewrite:
      uri: "/"
    route:
    - destination:
        host: strapi-backend
        port:
          number: 1337
```

We already tried nearly all permutations in rewriting, ENV URL/ADMIN URL setting (like ADMIN_URL=/backend/admin) etc., but the initial javascript is always requested from the root url.

**Comment by menflx** [comment]
We encountered the same problem when trying to host Strapi v5.0.3 under the subpath "/cms". Tried all possible combinations of absolute und relative URLs, but the admin page stays blank because the resources are still loaded from the root url.

`config/admin.ts`:
``` typescript
export default ({ env }) => ({
  auth: {
    secret: env('ADMIN_JWT_SECRET'),
  },
  apiToken: {
    salt: env('API_TOKEN_SALT'),
  },
  transfer: {
    token: {
      salt: env('TRANSFER_TOKEN_SALT'),
    },
  },
  url: env('ADMIN_URL'),
});

```

`config/server.ts`:
``` typescript
export default ({ env }) => ({
  host: env('HOST', '0.0.0.0'),
  port: env.int('PORT', 1337),
  app: {
    keys: env.array('APP_KEYS'),
  },
  url: env('PUBLIC_URL'),
  proxy: true,
  webhooks: {
    populateRelations: env.bool('WEBHOOKS_POPULATE_RELATIONS', false),
  },
});
```

For `ADMIN_URL`, we tried setting it to `/admin`, `/admin/`, `http://localhost:1337/cms/admin`, `http://localhost:1337/cms/admin/`.
For `PUBLIC_URL`, we tried setting it to `http://localhost:1337/cms` and `http://localhost:1337/cms/`.

**Comment by alexandrebodin** [comment]
Hi @menflx thanks for sharre your config I'll try to reproduce

**Comment by alexandrebodin** [comment]
We will review this community PR that seems to solve the issue https://github.com/strapi/strapi/pull/21507 as soon as we can

**Comment by mad-zephyr** [comment]
Hello and sorry, any updates related to this bug?

**Comment by alexandrebodin** [comment]
Hi @mad-zephyr  the PR is being reviewed right now. We are waiting for the author to comment back :) We will do some real world testing this week too

**Comment by mad-zephyr** [comment]
@alexandrebodin, Sorry again for my rush )) and yes I saw the PR related to this "bug" almost ready... Thanks for your time

**Comment by twistmachine** [comment]
hello everyone, i was wondering why my v4 configuration was not working in v5. i am waiting for your green light to test on my servers as well. thanks for your help

**Comment by EmmasBox** [comment]
I was running into this problem too. As a workaround I migrated to a subdomain, this is however extremely annoying as I have to create a page that can redirects people to the new CMS location.

**Comment by Matb85** [comment]
Hello, any updates on this? It is a dealbreaker for my team as we would like to host strapi on a subpath.

**Comment by Nijhazer** [comment]
There's a PR in the works to address this issue, and I've been keeping an eye on that. One question I have: In order to run Strapi on a subpath at present, you need a reverse proxy; _and_ if you're running Strapi in a Docker container, you have to provide a value for `BASE_URL` at container build time. Will the solution obviate the need for that?

**Comment by alexandrebodin** [comment]
Hello 👋  We have made an experimental release `0.0.0-experimental.25041a58cd2af628963c29620b4657416a4d772e` that you can test. It is aimed at being ISO with v4 (e.g requires a reverse-proxy). 

We have made a bunch of testing internally but we might have missed some edge cases. We would really appreciate if some of you can give it a try 🙏

**Comment by Nijhazer** [comment]
1. `npx create-strapi-app@0.0.0-experimental.25041a58cd2af628963c29620b4657416a4d772e` , etc
2. Update `config/server.ts` to set a value for `url`
3. `npm run develop`

Attempted values for `url`:
* `/strapi`: Launches `http://localhost:1337/strapi/admin` in a browser; 404
* `http://localhost:1337/strapi`: Launches `http://localhost:1337/strapi/admin` in a browser; 404

**Comment by alexandrebodin** [comment]
@Nijhazer what is your reverse proxy config ?

**Comment by alexandrebodin** [comment]
Sorry @Nijhazer We are not going to debug the docker image here :) this might be a img alpine issue missing a lib for sharp you will find a lot of resources for that outhere :) 

When you setup a reverse proxy in front of strapi you still need to setup the url in strapi so it generates the right urls in the admin assets, emails, etc

**Comment by Nijhazer** [comment]
Understandable. Outside of Docker, I performed these tests:

1. `npm run develop`, leave `URL` empty: Works
2. `npm run develop`, set `URL` to `/cms` or `http://localhost:1337/cms`: Fails; 404 when looking for the JavaScript bundle
3. Add `nginx` as a reverse proxy on `http://localhost:3000/cms`, leave `URL` empty: Fails; 404 when looking for the JavaScript bundle
4. Add `nginx` as a reverse proxy on `http://localhost:3000/cms`, set `URL` to `/cms` or `http://localhost:1337/cms`: Fails; 404 when looking for the JavaScript bundle

In summary, this build seems to behave no differently than current versions of Strapi. I'll leave it there so that others can report back with their own test results.

**Comment by alexandrebodin** [comment]
@Nijhazer  can you share your test nginx config ?

To anyone willing to test

- Make sure to set url in server.js to the public url (including the prefix) 
- Make sure to strip the prefix in your reverse proxy config as the api itself doesn't use the prefix

**Comment by Nijhazer** [comment]
Sure! More details...

I'm running `nginx` in a Docker container, and I'm running Strapi on the host. This is my `nginx` configuration:

```
server {
  listen 80;

  access_log /var/log/nginx/access.log;
  error_log /var/log/nginx/error.log;

  location /cms {
    proxy_pass http://host.docker.internal:3000/;
    rewrite ^/cms/(.*) /$1 break;

    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forwarded-Host $server_name;
  }
}
```

On the host, `nginx` is listening on port 4000, Strapi on port 3000.
When I start Strapi, `URL` is set to `http://localhost:4000/cms`

From there, I open my browser to `http://localhost:4000/cms`

Since my previous post, I've attempted a second test, so I'll explain the results below in light of each:

## Test 1: `npm run develop`

When I start Strapi using `npm run develop`, requests to `http://localhost:4000/cms/admin` fail:
* `http://localhost:4000/cms/admin/@vite/client` gets the HTML of the admin page in response
* `http://localhost:4000/cms/admin/.strapi/client/app.js` gets the HTML of the admin page in response

## Test 2: Create a `server.js` file

I create a `server.js` file with the following contents:
```
const strapi = require('@strapi/strapi');
strapi.createStrapi({ distDir: './dist' }).start();
```

Then I start Strapi:
```
pm2-runtime server.js
```

This time, everything works. Requests to `http://localhost:4000/cms/admin` load the admin panel as expected. Content Manager, etc seems to work fine.

Test 2 is effectively the same solution that I'm using right now with Strapi v4. The main issue with it is that changes to `URL` require rebuilding the app, which prevents deploying a container in more than one environment (i.e., a separate staging and production site). So, as an additional test, I changed `URL` to `http://localhost:5000/cms`, updated the `nginx` config, restarted Strapi and `nginx`, and visited `http://localhost:5000/cms/admin` in a browser. That produces a 404, with this message in the Strapi log:
```
Error: ENOENT: no such file or directory, open '/Users/<me>/<project-dir>/node_modules/@strapi/admin/build/index.html'
```

**Comment by alexandrebodin** [comment]
Thanks @Nijhazer I'm not sure how your test 1 fails vs test 2 from reading your message 🤔 if you have a repo to checkout and test that could be very helpfull. In the meatime this seems to confirm v5 & v4 are now ISO.

If we don't get more tests / comments we will release the current work for next week.

About the admin needing a rebuild indeed, once the admin code is built it includes certain fixed variables. I'll see if we can make it work with relative URLs when admin & api are on the same domain.

**Comment by alexandrebodin** [comment]
Hi, FYI we released a new experimental version to test that adds better support for relative paths

`0.0.0-experimental.bec5a58066c034a7ebf5b14df62560e68a456fa3`

**Comment by yfs-2000** [comment]
I have successfully configured the subpath for Strapi.
```
config/admin  url: /strapi-admin/admin
config/server   url: http://localhost:1400/strapi-admin/api

nginx
server {
        # Listen HTTPS
        listen 1400;
        server_name localhost;
        location /strapi-admin/api/ {
            proxy_pass http://localhost:1340/;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection 'upgrade';
            proxy_set_header Host $host;
            proxy_cache_bypass $http_upgrade;
        }
        location /strapi-admin/admin/ {
           
            proxy_pass http://localhost:1340/strapi-admin/admin/;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection 'upgrade';
            proxy_set_header Host $host;
            proxy_cache_bypass $http_upgrade;
        }

    }

```

**Comment by Nijhazer** [comment]
Returning from holiday...

Using the new build `0.0.0-experimental.bec5a58066c034a7ebf5b14df62560e68a456fa3`, I repeated [Test 2](https://github.com/strapi/strapi/issues/21184#issuecomment-2494530036), only this time I changed `URL` to `/cms` instead of `http://localhost:5000/cms` . And it works! Amazing!

**Comment by CharuGoyalS** [comment]
Thanks for sharing the fix ! 

When are we planning to integrate this fix to Strapi stable version?

**Comment by alexandrebodin** [comment]
Looks like we are ready to do a last review and merge so that should be release next week in the upcoming patch release 👍

**Comment by wow64cpu** [comment]
> Returning from holiday...
> 
> Using the new build `0.0.0-experimental.bec5a58066c034a7ebf5b14df62560e68a456fa3`, I repeated [Test 2](https://github.com/strapi/strapi/issues/21184#issuecomment-2494530036), only this time I changed `URL` to `/cms` instead of `http://localhost:5000/cms` . And it works! Amazing!

Hi, I also have a problem with deploying strapi on a custom route

My domain contains a subdomain and I try to place strapi on the `/cms` path, for example - `something.example.com/cms`, respectively, the admin panel should be available at the URL `something.example.com/cms/admin`

But it does not work

I see that when trying to go to the specified page, the browser tries to request scripts at the wrong path, for example - `/admin`

I run strapi in docker and from the outside I proxy requests to it through nginx

![image_2024-12-06_14-10-29](https://github.com/user-attachments/assets/2de3bdb3-607e-42e2-abbc-f02560ec763b)

**Build `0.0.0-experimental.bec5a58066c034a7ebf5b14df62560e68a456fa3` also working for me**

**Comment by remi-viau** [comment]
Hello all,

Im still unable to make it work in v5 with simple setup for subpath i used in v4. 

server.ts with public URL https://whatever.com**/cms**:

`export default ({ env }) => ({
  host: env("HOST", "0.0.0.0"),
  port: env.int("PORT", 1337),
  url: env('PUBLIC_URL'),
  proxy: true,
  app: {
    keys: env.array("APP_KEYS"),
  },
  webhook: {
    populateRelations: false,
  },
});`

Proxy side (haproxy) : 

`backend service_strapi

    option redispatch
    option abortonclose
    option forwardfor
    http-request set-path "%[path,regsub(^/cms/?,/)]"
    server strapi pod0-ca-east-1:4002 check`

Working setup : 

server.ts with public URL https://cms.whatever.com:

`export default ({ env }) => ({
  host: env("HOST", "0.0.0.0"),
  port: env.int("PORT", 1337),
  url: env('PUBLIC_URL'),
  proxy: true,
  app: {
    keys: env.array("APP_KEYS"),
  },
  webhook: {
    populateRelations: false,
  },
});`

Proxy side (haproxy) : 

`backend service_strapi

    option redispatch
    option abortonclose
    option forwardfor
    server strapi pod0-ca-east-1:4002 check`


Im on 5.6.0
The subpath version seems to work as i receive HTTP 200 : 

[2024-12-19 14:20:24.145] http: GET /admin/@vite/client (8 ms) 200
[2024-12-19 14:20:24.147] http: GET /admin/.strapi/client/app.js (7 ms) 200
[2024-12-19 14:20:24.149] http: GET /admin/@react-refresh (8 ms) 200

 in console but Chrome and firefox return in console : 

client:1 Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "text/html". Strict MIME type checking is enforced for module scripts per HTML spec.Understand this errorAI
app.js:1 Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "text/html". Strict MIME type checking is enforced for module scripts per HTML spec.Understand this errorAI
@react-refresh:1 Failed to load module script: Expected a JavaScript module script but the server responded with a MIME type of "text/html". Strict MIME type checking is enforced for module scripts per HTML spec.

**Comment by remi-viau** [comment]
> @remi-viau What version of Strapi are you using ?

latest 5.6.0

**Comment by remi-viau** [comment]
> @remi-viau What version of Strapi are you using ?

After checking deeper, i see that the file returned for @vite in subpath is : <!doctype html>
```html
<html lang="en">
  <!--
This file was automatically generated by Strapi.
Any modifications made will be discarded.
-->
  <head>
    <script type="module">import { injectIntoGlobalHook } from "/cms/admin/@react-refresh";
injectIntoGlobalHook(window);
window.$RefreshReg$ = () => {};
window.$RefreshSig$ = () => (type) => type;</script>

    <script type="module" src="/cms/admin/@vite/client"></script>

    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1, viewport-fit=cover"
    />
    <meta name="robots" content="noindex" />
    <meta name="referrer" content="same-origin" />
    <title>Strapi Admin</title>
    <style>
      html,
      body,
      #strapi {
        height: 100%;
      }
      body {
        margin: 0;
        -webkit-font-smoothing: antialiased;
      }
    </style>
  </head>
  <body>
    <div id="strapi"></div>
    <noscript
      ><div class="strapi--root">
        <div class="strapi--no-js">
          <style type="text/css">
            .strapi--root {
              position: absolute;
              top: 0;
              right: 0;
              left: 0;
              bottom: 0;
              background: #fff;
            }

            .strapi--no-js {
              position: absolute;
              top: 50%;
              left: 50%;
              transform: translate(-50%, -50%);
              text-align: center;
              font-family: helvetica, arial, sans-serif;
            }
          </style>
          <h1>JavaScript disabled</h1>
          <p>
            Please
            <a href="https://www.enable-javascript.com/">enable JavaScript</a>
            in your browser and reload the page to proceed.
          </p>
        </div>
      </div></noscript
    >
    <script type="module" src="/cms/admin/.strapi/client/app.js"></script>
  </body>
</html>
```

**Comment by alexandrebodin** [comment]
can you share the exact config you have in server.js & admin.js. I'm not using haproxy often so won't be able to help you on that side though.

**Comment by theCompanyDream** [comment]
please do I'm having trouble getting this to work in my local

**Comment by remi-viau** [comment]
Hello @alexandrebodin 

For your information i encounter the same problem when i bypass the proxy
i tried with PUBLIC_URL :

PUBLIC_URL=http://domain.com/cms
PUBLIC_URL=/cms

Same behaviour, file sended by the server (client and app.js)  are not the one in place in the folder of the server and show to enable js. Its better than before but still not working. I use this proxy+config in v4 without any issue

here my conf : 

server.ts
```javascript
export default ({ env }) => ({
  host: env("HOST", "0.0.0.0"),
  port: env.int("PORT", 1337),
  url: env("PUBLIC_URL"),
  proxy: env.bool("IS_PROXIED", true),
  app: {
    keys: env.array("APP_KEYS"),
  },
  webhook: {
    populateRelations: false,
  },
});
```
admin.ts
```javascript
export default ({ env }) => ({
  auth: {
    secret: env('ADMIN_JWT_SECRET'),
  },
  apiToken: {
    salt: env('API_TOKEN_SALT'),
  },
  transfer: {
    token: {
      salt: env('TRANSFER_TOKEN_SALT'),
    },
  },
  flags: {
    nps: env.bool('FLAG_NPS', true),
    promoteEE: env.bool('FLAG_PROMOTE_EE', true),
  },
});
```

**Comment by rafaelklaessen** [comment]
@remi-viau When attempting to migrate my project to Strapi v5, I run into the exact same issue where the admin panel at a subpath cannot load any JS resources since those resources simply return the Vite's internal index.html file. We are using nginx, so it's definitely not a haproxy issue.

**Comment by itstunahan** [comment]
The same problem still persists, has anyone found a solution?

**Comment by alexandrebodin** [comment]
Hello @remi-viau thanks for the details. I'll try to reproduce this in the next few days. Catching up after end of year break. 

For everyone else, Can you please provide both the server, admin configs and reverse proxy configs so we have complete reproduction steps 🙏

**Comment by alexandrebodin** [comment]
FYI I identified a pbl in dev mode with the vite middleware not knowing what path to look for with the proxying. It's not going to be trivial to fix.

Here is an example config for haproxy that will work in prod when building & serving so you can at least deploy how you want to in prod

`config/server.js`
```js
module.exports = ({ env }) => ({
  host: env('HOST', '0.0.0.0'),
  port: env.int('PORT', 1337),
  url: 'http://localhost:5000/cms',
});
```
`config/admin.js`
```js
// no url 
```

`haproxy.cfg`
```
frontend http_in
    bind *:5000
    mode http
    http-request set-path %[path,regsub(^/cms/,/)]

    http-request set-header X-Forwarded-For %[src]
    http-request set-header X-Forwarded-Host %[req.hdr(Host)]

    default_backend backend

backend backend
    mode http
    server backend_server localhost:1337
```

**Comment by alexandrebodin** [comment]
I released this experimental release `0.0.0-experimental.c2680dea29af0c9e9a69c0474b40c8c2ab6f70ac` that should address the dev mode issue I mentionned earlier. (Tested it with the setup I shared above) Can anyone try that out please 🙏

**Comment by jackbuehner** [comment]
@alexandrebodin I can confirm that your change addresses the dev mode issue. Thank you for working on this! I tested it with 

`config/server.ts`
```
export default ({ env }) => ({
  host: env('HOST', '0.0.0.0'),
  port: env.int('PORT', 1337),
  url: 'http://localhost:3000/strapi',
});
```

`config/admin.ts`
```
export default ({ env }) => ({
  url: '/cms',
});
```

I am weird and reverse proxy everything by intercepting requests starting with `/strapi` in my through my SvelteKit `hooks.server.ts` rather than using something like nginx, but the effect is similar. I do this so that my main website and my admin panel are on the same origin without needing to put my SvelteKit deployment behind a reverse proxy too. But if you think it would be helpful for me to share an example, I am happy to create a version of my `hooks.server.ts` with only the relevant code.

<details>
<summary>How I did this from Strapi v5-v5.4</summary>

On the strapi server, I had this `src/admin/vite.config.ts`
```
import { defineConfig, mergeConfig, type UserConfig } from 'vite';

export default (config: UserConfig) => {
  const env = config.define!['process.env'];

  const strapiBackendPathname = new URL(env.STRAPI_ADMIN_BACKEND_URL).pathname;

  return mergeConfig(
    config,
    defineConfig({
      base: `${strapiBackendPathname}${config.base || ''}`,
      define: {
        'process.env': {
          ...env,
          ADMIN_PATH: `${strapiBackendPathname}${env.ADMIN_PATH}`,
        },
      },
    })
  );
};
```

And in my reverse proxy, whenever vite returned HTML, I would replace `<script type="module" src="/cms/strapi-` in the HTML with `<script type="module" src="/strapi/cms/strapi-` so that the main script that vite needs in order to start could be accessed. With the experimental release with the change from PR #22579, these are no longer necessary!

</details>

**Comment by troyere** [comment]
I have updated from v4 to v5.21.0, and I have exactly the same problem : the subpath config does not work at all anymore

**Comment by sanjuant** [comment]
Hello, i have the same problem 👍 

<img width="1611" height="649" alt="Image" src="https://github.com/user-attachments/assets/ae2bd146-96ee-4ef2-b8eb-180923759967" />

**Comment by Tofandel** [comment]
Why is this issue closed?

This is still problematic.

 I'm trying to serve the admin panel via passenger in a subpath like my.app/dashboard, I tried setting the admin.url config to '/dashboard' but I get a 404 from strapi when I go there.


Also if I try to edit the vite.config.ts and change the `base` option of vite, strapi still overrides it later.

I think we should be able to set a global prefix as an option here https://github.com/strapi/strapi/blob/4828e44800fa93cb20b61e4256f9362eb94762bc/packages/core/core/src/services/server/index.ts#L26

Or here for just the dashboard

https://github.com/strapi/strapi/blob/4828e44800fa93cb20b61e4256f9362eb94762bc/packages/core/core/src/services/server/admin-api.ts#L6

But it's hardcoded to nothing right now so the admin path config is completely ignored by koa

I just tried and if I hotfix and replace prefix: '' with `strapi.config.get('admin.path', '')`, the app starts serving properly.. With the only exception being a hardcoded js redirect, so when you go directly to the url you get redirected to the wrong login url

https://github.com/strapi/strapi/blob/4828e44800fa93cb20b61e4256f9362eb94762bc/packages/core/admin/admin/src/core/store/configure.ts#L95

(Should be `/dashboard/auth/login` in my case)
